util/progress/progressui/display.go

```diff
@@ -69,53 +88,208 @@ func DisplaySolveStatus(ctx context.Context, c console.Console, w io.Writer, ch
+func NewConsoleDisplay(c console.Console, opts ...DisplaySolveStatusOpt) SolveStatusDisplay {
+	dsso := newDisplaySolveStatusOpts(opts...)
+	if dsso.phase == "" {
+		dsso.phase = "Building"
+	}
+	return SolveStatusDisplay{
+		disp: &consoleDisplay{
+			t:    newTrace(c, true),
+			disp: &display{c: c, phase: dsso.phase, desc: dsso.consoleDesc},
+		},
+	}
+}
+
+func (d *consoleDisplay) init(displayTimeout time.Duration) {
+	d.displayLimiter = rate.NewLimiter(rate.Every(displayTimeout), 1)
+}
+
+func (d *consoleDisplay) update(ss *client.SolveStatus) {
+	d.width, d.height = d.disp.getSize()
+	d.t.update(ss, d.width)
+	if !d.displayLimiter.Allow() {
+		// Exit early as we are not allowed to update the display.
+		return
+	}
+	d.refresh()
+}
+
+func (d *consoleDisplay) refresh() {
+	d.disp.print(d.t.displayInfo(), d.width, d.height, false)
+}
+
+func (d *consoleDisplay) done() {
+	d.width, d.height = d.disp.getSize()
+	d.disp.print(d.t.displayInfo(), d.width, d.height, true)
+	d.t.printErrorLogs(d.t.w)
+}
+
+type plainDisplay struct {
+	t              *trace
+	printer        *textMux
+	displayLimiter *rate.Limiter
+}
+
+// NewPlainDisplay creates a new SolveStatusDisplay that outputs the status
+// in a human-readable plain-text format.
+func NewPlainDisplay(w io.Writer, opts ...DisplaySolveStatusOpt) SolveStatusDisplay {
+	dsso := newDisplaySolveStatusOpts(opts...)
+	return SolveStatusDisplay{
+		disp: &plainDisplay{
+			t: newTrace(w, false),
+			printer: &textMux{
+				w:    w,
+				desc: dsso.textDesc,
+			},
+		},
+	}
+}
+
+func (d *plainDisplay) init(displayTimeout time.Duration) {
+	d.displayLimiter = rate.NewLimiter(rate.Every(displayTimeout), 1)
+}
+
+func (d *plainDisplay) update(ss *client.SolveStatus) {
+	if ss != nil {
+		d.t.update(ss, 80)
+		if !d.displayLimiter.Allow() {
+			// Exit early as we are not allowed to update the display.
+			return
 		}
 	}
+	d.refresh()
+}
+
+func (d *plainDisplay) refresh() {
+	d.printer.print(d.t)
+}
+
+func (d *plainDisplay) done() {
+	// Force the display to refresh.
+	d.refresh()
+	// Print error logs.
+	d.t.printErrorLogs(d.t.w)
+}
+
+type rawJSONDisplay struct {
+	enc *json.Encoder
+	w   io.Writer
+}
+
+// NewRawJSONDisplay creates a new SolveStatusDisplay that outputs an unbuffered
+// output of status update events.
+func NewRawJSONDisplay(w io.Writer, opts ...DisplaySolveStatusOpt) SolveStatusDisplay {
+	enc := json.NewEncoder(w)
+	enc.SetIndent("", "  ")
+	return SolveStatusDisplay{
+		disp: &rawJSONDisplay{
+			enc: enc,
+			w:   w,
+		},
+	}
+}
+
+func (d *rawJSONDisplay) init(displayTimeout time.Duration) {
+	// Initialization parameters are ignored for this display.
+}
+
+func (d *rawJSONDisplay) update(ss *client.SolveStatus) {
+	_ = d.enc.Encode(ss)
+}
+
+func (d *rawJSONDisplay) refresh() {
+	// Unbuffered display doesn't have anything to refresh.
+}
+
+func (d *rawJSONDisplay) done() {
+	// Flush if the underling writer supports that method.
+	if flusher, ok := d.w.(interface {
```

Comment:
I'm not sure if this should be handled here. If someone passes a buffered writer it is their responsibility to flush it after the display has completed.

Reply:
I'll remove this. My intention with this was to avoid any needed flushing for things like `Stdout` and `Stderr` and just to provide some convenience and parity to how the TTY output functions, but I don't think this is really necessary.